File: Behavioral/Visitor/python/visitor.py
from abc import ABC, abstractmethod


class Element(ABC):

    @abstractmethod
    def accept(self, visitor):
        pass


class Visitor(ABC):

    @abstractmethod
    def visist_a(self, element_a):
        pass

    @abstractmethod
    def visist_b(self, element_b):
        pass


class ElementA(Element):

    def accept(self, visitor: Visitor):
        visitor.visist_a(self)

    def feature_a(self):
        return "Feature of Element A"


class ElementB(Element):

    def accept(self, visitor: Visitor):
        visitor.visist_b(self)

    def feature_b(self):
        return "Feature of Element B"


class ConcreteVisitorAlpha(Visitor):

    def visist_a(self, element_a: ElementA):
        print("Concrete Visitor Alpha visiting %s" % element_a.feature_a())

    def visist_b(self, element_b: ElementB):
        print("Concrete Visitor Alpha visiting %s" % element_b.feature_b())


class ConcreteVisitorBeta(Visitor):

    def visist_a(self, element_a: ElementA):
        print("Concrete Visitor Beta visiting %s" % element_a.feature_a())

    def visist_b(self, element_b: ElementB):
        print("Concrete Visitor Beta visiting %s" % element_b.feature_b())


def main():
    elements = [ElementA(), ElementB()]
    visitors = [ConcreteVisitorAlpha(), ConcreteVisitorBeta()]

    for element in elements:
        for visitor in visitors:
            element.accept(visitor)


if __name__ == "__main__":
    main()

File: Behavioral/Visitor/scala/Visitor.scala

trait Element {
  def accept(visitor: Visitor)
}

trait Visitor {

  def visitElementA(elementA: ElementA)

  def visitElementB(elementB: ElementB)
}


class ElementA extends Element {

  override def accept(visitor: Visitor): Unit = {
    visitor.visitElementA(this)
  }

  def featureOfElementA() = "Feature of element A"

}

class ElementB extends Element {

  override def accept(visitor: Visitor): Unit = {
    visitor.visitElementB(this)
  }

  def featureOfElementB() = "Feature of element B"

}

class ConcreteVisitorAlpha extends Visitor {

  override def visitElementA(elementA: ElementA): Unit = {
    println(f"Concrete Visitor Alpha visiting ${elementA.featureOfElementA()}")
  }

  override def visitElementB(elementB: ElementB): Unit = {
    println(f"Concrete Visitor Alpha visiting ${elementB.featureOfElementB()}")
  }
}

class ConcreteVisitorBeta extends Visitor {

  override def visitElementA(elementA: ElementA): Unit = {
    println(f"Concrete Visitor Beta visiting ${elementA.featureOfElementA()}")
  }

  override def visitElementB(elementB: ElementB): Unit = {
    println(f"Concrete Visitor Beta visiting ${elementB.featureOfElementB()}")
  }
}


object VisitorPattern {
  def main(args: Array[String]): Unit = {
    val elements = Seq(new ElementA(), new ElementB())
    val visitors = Seq(new ConcreteVisitorAlpha(), new ConcreteVisitorBeta())

    elements.foreach(element => {
      visitors.foreach(visitor => {
        element.accept(visitor)
      })
    })
  }
}
